src/gluonts/model/wavenet/_network.py

```diff
@@ -547,59 +586,204 @@ def blow_up(u):
+        past_target
+            Past target: (batch_size, receptive_field)
+        past_observed_values
+            Observed value indicator for the past target: (batch_size, receptive_field)
+        past_time_feat
+            Past time features: (batch_size, num_time_features, receptive_field)
+        future_time_feat
+            Future time features: (batch_size, num_time_features, pred_length)
+        scale
+            scale of the time series: (batch_size, 1)
+
+        Returns
+        -------
+        Tensor
+            Prediction samples with shape (batch_size, num_samples, pred_length)
+        """
+
+        samples_bin_ix = self.predict(
+            F,
+            feat_static_cat=feat_static_cat,
+            past_target=past_target,
+            past_observed_values=past_observed_values,
+            past_time_feat=past_time_feat,
+            future_time_feat=future_time_feat,
+            scale=scale,
+            num_samples=self.num_samples,
+            sample_based_forecast=True,
         )
-        samples = self.post_transform(samples)
+        samples = self.post_transform(samples_bin_ix)
         samples = F.broadcast_mul(scale.expand_dims(axis=1), samples)
         return samples
+
+
+class WaveNetDistributionPredictor(WaveNetPredictor):
+    """
+    Runs Wavenet generation in an auto-regressive manner using caching for
+    speedup [PKC+16]_ to produce (parametric) distribution forecast.
+
+    Same arguments as `WaveNetPredictor`.
+    """
+
+    def hybrid_forward(
+        self,
+        F,
+        feat_static_cat: Tensor,
+        past_target: Tensor,
+        past_observed_values: Tensor,
+        past_time_feat: Tensor,
+        future_time_feat: Tensor,
+        scale: Tensor,
+    ) -> Tuple[Tuple[Tensor, Tensor], Tensor]:
+        """
+        Computes the forecast distribution from the trained wavenet model.
+
+        Parameters
+        ----------
+        F
+        feat_static_cat
+            Static categorical features: (batch_size, num_cat_features)
+        past_target
+            Past target: (batch_size, receptive_field)
+        past_observed_values
+            Observed value indicator for the past target: (batch_size, receptive_field)
+        past_time_feat
+            Past time features: (batch_size, num_time_features, receptive_field)
+        future_time_feat
+            Future time features: (batch_size, num_time_features, pred_length)
+        scale
+            scale of the time series: (batch_size, 1)
+
+        Returns
+        -------
+        Tuple containing `dist_args`, `loc` and `scale`.
+
+            `dist_args` is a pair of tensors containing logarithm of bin probabilities with shape
+            (batch_size, pred_length, num_bins) and bin values with shape (num_bins,).
+
+            `loc` is a Tensor with shape (batch_size, 1).
+
+            `scale` is a Tensor with shape (batch_size, 1).
+        """
+
+        bin_probs = self.predict(
```

Comment:
Using a single sample for the autoregressive part here is probably not ideal. 
Two ideas to improve this:

- pass the temperature for drawing samples. Then we could run this with temperature 0 picking the bin with max probability in each step
- run the prediction with more than one sample and consolidate afterwards, by averaging the bin probabilities over the samples. 

The second option is the correct way, but we have to draw more samples. The first is not really correct but may work in practice.

Reply:
Mmh, I think this is not what you want. We probably want to use the T=0 for drawing the samples, but T=1 for the distribution output, right?

Reply:
Yes, that is why the [returned `bin_probs`](https://github.com/awslabs/gluon-ts/pull/773/files/3fa3f5c6929af57049ef037d5098fdcd79afebcd#diff-0ca3301b70787528429402b5e0015586R514) are not affected by temperature (T=1 while computing these probs). I also added a [note](https://github.com/awslabs/gluon-ts/pull/773/files/3fa3f5c6929af57049ef037d5098fdcd79afebcd#diff-0ca3301b70787528429402b5e0015586R501) clarifying this.

Reply:
Ok, makes sense.